x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/boxplot/InternalBoxplot.java

```diff
@@ -0,0 +1,200 @@
+        this.state = state;
+        this.format = formatter;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public InternalBoxplot(StreamInput in) throws IOException {
+        super(in);
+        format = in.readNamedWriteable(DocValueFormat.class);
+        state = TDigestState.read(in);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(format);
+        TDigestState.write(state, out);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return BoxplotAggregationBuilder.NAME;
+    }
+
+    @Override
+    public double getMin() {
+        return state.getMin();
+    }
+
+    @Override
+    public double getMax() {
+        return state.getMax();
+    }
+
+    @Override
+    public double getQ1() {
+        return state.quantile(0.25);
+    }
+
+    @Override
+    public double getQ2() {
+        return state.quantile(0.5);
+    }
+
+    @Override
+    public double getQ3() {
+        return state.quantile(0.75);
+    }
+
+    @Override
+    public String getMinAsString() {
+        return valueAsString(Metrics.min.name());
+    }
+
+    @Override
+    public String getMaxAsString() {
+        return valueAsString(Metrics.max.name());
+    }
+
+    @Override
+    public String getQ1AsString() {
+        return valueAsString(Metrics.q1.name());
+    }
+
+    @Override
+    public String getQ2AsString() {
+        return valueAsString(Metrics.q2.name());
+    }
+
+    @Override
+    public String getQ3AsString() {
+        return valueAsString(Metrics.q3.name());
+    }
+
+    @Override
+    public double value(String name) {
+        Metrics metrics = Metrics.valueOf(name);
+        switch (metrics) {
+            case min:
+                return getMin();
+            case max:
+                return getMax();
+            case q1:
+                return getQ1();
+            case q2:
+                return getQ2();
+            case q3:
+                return getQ3();
+            default:
+                throw new IllegalArgumentException("Unknown value [" + name + "] in common stats aggregation");
+        }
+    }
+
+    // for testing only
+    DocValueFormat format() {
+        return format;
+    }
+
+    // for testing only
+    TDigestState state() {
+        return state;
+    }
+
+    @Override
+    public InternalBoxplot reduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        TDigestState merged = null;
+        for (InternalAggregation aggregation : aggregations) {
+            final InternalBoxplot percentiles = (InternalBoxplot) aggregation;
+            if (merged == null) {
+                merged = new TDigestState(percentiles.state.compression());
+            }
+            merged.add(percentiles.state);
+        }
+        return new InternalBoxplot(name, merged, format, pipelineAggregators(), metaData);
+    }
+
+    static class Fields {
```

Comment:
I *thought* that these `Fields` objects had gone out of favor. But I'm not really in the loop.